### PR TITLE
Data Attributes

### PR DESCRIPTION
**Pull request by Denommus**
## Description

- Introduce a migration to have data attributes in the chain
- Two attributes introduced
- Computing the new attributes

## Related Issue

Refs liftedinit/roadmap#5

## Testing

Added a new unit test that checks if the migration mechanism works and if the data attributes are being correctly computed

## Breaking Changes (if applicable)

Changes the consensus with a new migration

## Checklist:

- [X] I have read and followed the CONTRIBUTING guidelines for this project.
- [X] I have added or updated tests and they pass.
- [X] I have added or updated documentation and it is accurate.
- [X] I have noted any breaking changes in this module or downstream modules.

### CODE REVIEW

File: src/many-ledger/src/storage.rs
Comment (fmorency) [inline src/many-ledger/src/storage.rs:678]: make a constant with this key

File: src/many-ledger/src/storage.rs
Comment (fmorency) [inline src/many-ledger/src/storage.rs:618]: Not a fan of this

**Review by fmorency — CHANGES_REQUESTED** [review]
Some comments
- Why not re-use/adapt the existing migration module?
- Migration code should be isolated
- Ideally, post-migration code path should be isolated as well

It would be nice not to have to modify `storage.rs` to add a new migration. However, that goal might be out of scope of this task.

I would mostly try to reduce coupling in what you did.

I would have to think more about it, but are we 110% sure we can't have any race condition, e.g., in `count_addresses` ?

File: tests/resiliency/ledger/migration.bats
Comment (fmorency) [inline tests/resiliency/ledger/migration.bats:38]: Rename the test to `$SUITE: data attribute migration works`

File: tests/resiliency/ledger/migrations.toml
Comment (fmorency) [inline tests/resiliency/ledger/migrations.toml:2]: Test the `issue` stuff as well

**Review by fmorency — CHANGES_REQUESTED** [review]
I think the design could be better and more straightforward.

We should probably **version** the migrations as well. A migration code could be updated, and we need to track that. We could even version & hash if we feel like it.

- I would like to see a consistency test where you test for a race condition, i.e., test that the `count ± 1` stays coherent and that you're not updating an old value

File: src/many-ledger/src/migration.rs
Comment (Denommus) [reply src/many-ledger/src/migration.rs:1]: Good one

File: src/many-ledger/src/module.rs
Comment (Denommus) [reply src/many-ledger/src/module.rs:39]: Absent mindness when creating migrstion::data. Will fix

File: src/many-ledger/tests/migration.rs
Comment (Denommus) [reply src/many-ledger/tests/migration.rs:39]: I'm only importing the correct trait.

File: src/many-ledger/src/storage.rs
Comment (Denommus) [reply src/many-ledger/src/storage.rs:376]: It shouldn't run the active_migrations again, it should migrate only for migrations that still aren't on active_migrations. I can add a test for that, but this is not an inconsistency per se.

File: src/many-ledger/src/migration.rs
Comment (fmorency) [reply src/many-ledger/src/migration.rs:1]: This is an example of what I had in mind: https://www.rustexplorer.com/b/cq68l8

It makes it very easy to add new types of migrations with minimal code change, i.e., no `MigrationName` to track, etc.

**Review by fmorency — CHANGES_REQUESTED** [review]
Some design hints.

You did not address a lot of the comments from my preview review.

File: src/many-ledger/src/storage.rs
Comment (fmorency) [reply src/many-ledger/src/storage.rs:1569]: You could add a new `storage/migration/data` module.

I did something similar to this in `many-kvstore` for `events` and `account`, check it out.

The file `src/storage.rs` is too big.

File: src/many-ledger/src/migration.rs
Comment (fmorency) [inline src/many-ledger/src/migration.rs:73]: Why not write to `MIGRATIONS_KEY` only once with the final `active_migrations`?

**Review by fmorency — CHANGES_REQUESTED** [review]
Thanks @Denommus. 

The design is much better and I think it's good enough for now. We can improve on it later on.

Some nits and discussions, but nothing major.

Edit: Oh, and CI is broken, and there are conflicts.

@hansl I think you can review this now.

**Comment by fmorency** [comment]
@Denommus: I would also like to stress test this to ensure we don't get a data race regression.

File: src/many-ledger/src/module.rs
Comment (Denommus) [reply src/many-ledger/src/module.rs:840]: First, this is not a "migration", this is just an additional module, modules do not migrate. Second, it would not be possible to put it in a separate module because `DataModuleBackend` is defined externally, so the trait impl needs to be in the same module as the `LedgerModuleImpl` definition.

File: src/many-ledger/src/storage.rs
Comment (Denommus) [reply src/many-ledger/src/storage.rs:747]: I don't understand your point. This update needs to happen in a specific position of the storage code, and it cannot be generic because other migrations might have hooks in different positions. So it needs to be data-attribute specific.

File: src/many-ledger/src/migration.rs
Comment (fmorency) [reply src/many-ledger/src/migration.rs:58]: A `BTreeSet` is automatically sorted. You could use that and ditch the `sort_by`.

File: src/many-ledger/src/module.rs
Comment (fmorency) [reply src/many-ledger/src/module.rs:840]: Sorry, you could do a `module/data.rs` or the like and place the data-related impl for the LedgerModuleImpl in it.

I'm doing that in `kvstore` to lighten `module.rs`, e.g., https://github.com/liftedinit/many-framework/blob/2b8bde732618f9342f2f4aeb367d69a250057862/src/many-kvstore/src/module/account.rs#L119

File: src/many-ledger/src/migration.rs
Comment (Denommus) [reply src/many-ledger/src/migration.rs:58]: I understand it now. I think it would be too verbose for too little of a benefit.

File: src/many-ledger/src/storage.rs
Comment (fmorency) [reply src/many-ledger/src/storage.rs:741]: We can improve this, later on, I won't block the merge.

File: src/many-ledger/src/storage.rs
Comment (fmorency) [inline src/many-ledger/src/storage.rs:586]: Can you comment on why this is needed?

**Review by fmorency — COMMENTED** [review]
Thanks!

Small stuff.

@hansl you're up!

**Review by hansl — CHANGES_REQUESTED** [review]
By default, migrations should be enabled but not migrates. If I start a new network, it should calculate the number of accounts, for example. But if I use it in an existing network, it should apply a migration at a certain block height. This maybe done separately, but I will want to discuss with Felix before we approve this PR.

Additionally, here's a diff for changes that make the docker work again. Otherwise it tries to read an empty JSON file which is invalid.

```diff
diff --git a/docker/e2e/Makefile.ledger b/docker/e2e/Makefile.ledger
index 98040a6..e620fc5 100644
--- a/docker/e2e/Makefile.ledger
+++ b/docker/e2e/Makefile.ledger
@@ -37,7 +37,7 @@ genfiles-ledger/docker-compose.json: genfiles-common/jsonnet-docker docker-compo
 		--tla-code abci_tag=\"$(ABCI_TAG)\" \
 		--tla-code ledger_tag=\"$(LEDGER_TAG)\" \
 		--tla-code tendermint_tag=\"$(TENDERMINT_VERSION)\" \
-		--tla-code migrations=\"/genfiles/migrations.toml\" \
+		--tla-code migrations=\"/genfiles/migrations.json\" \
 		--tla-code allow_addrs=$(ALLOW_ADDRS) \
 		-o /$@
 
@@ -53,11 +53,11 @@ genfiles-ledger/generate-tendermint-e2e-config:
 
 genfiles-ledger/node%: genfiles-common/openssl-docker genfiles-common/tendermint-docker
 	mkdir -p "$@/tendermint"
-	rm -f "$@/migrations.toml"
+	rm -f "$@/migrations.json"
 ifdef MIGRATIONS
-	cp $(MIGRATIONS) "$@/migrations.toml"
+	cp $(MIGRATIONS) "$@/migrations.json"
 else
-	touch "$@/migrations.toml"
+	echo '[]' > "$@/migrations.json"
 endif
 	docker run --user $$(id -u) -it --rm -v "${PWD}/$@/tendermint:/tendermint" tendermint/tendermint:v$(TENDERMINT_VERSION) init validator
 	docker run --user $$(id -u) -it --rm -v ${PWD}/$@/:/export alpine/openssl genpkey -algorithm Ed25519 -out /export/ledger.pem
```

**Review by hansl — APPROVED** [review]
Approving as is, but there's still work to be done here. We can take it from here (documentation etc). Thanks Yuri!

**Pull request merged**
